📖  doc: Fix multigroup migration guide for v4

juanluisvaladas opened the pull request:
<!--

Hiya!  Welcome to Kubebuilder!  For a smooth PR process, please ensure
that you include the following information:

* a description of the change
* the motivation for the change
* what issue it fixes, if any, in GitHub syntax (e.g. Fixes #XYZ)

Both the description and motivation may reference other issues and PRs,
but should be mostly understandable without following the links (e.g. when
reading the git commit log).

Please don't @-mention people in PR or commit messages (do so in an
additional comment).

please add an icon to the title of this PR depending on the type:

- ⚠ (:warning:): breaking
- ✨ (:sparkles:): new non-breaking feature
- 🐛 (:bug:): bugfix
- 📖 (:book:): documentation
- 🌱 (:seedling:): infrastructure/other

See https://sigs.k8s.io/kubebuilder-release-tools for more information.

**PLEASE REMOVE THIS COMMENT BLOCK BEFORE SUBMITTING THE PR** (the bits
between the arrows)

-->

Fixes #3401 .
This is necessary because the paths of the warning were incorrect.

I added two separate instructions for both v3 and v4 as there were a few more differences besides the path.

varshaprasad96 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -104,3 +103,92 @@ single-group projects).
+
+
+Next, we'll need to update all the references to the old package name. 
+For CronJob, that'll be `main.go` and `controllers/batch/cronjob_controller.go`. 
```
```suggestion
For CronJob, the paths to be replaced would be `main.go` and `controllers/batch/cronjob_controller.go` to their respective locations in the new project structure. 
```

varshaprasad96 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -104,3 +103,92 @@ single-group projects).
+
+Finally, fix the PROJECT file manually, the command `kubebuilder edit --multigroup=true`
+sets our project to multigroup, but it doesn't fix the path of the existing APIs.
+For each resource need to modify the path and perform no additional modifications.
```
```suggestion
For each resource we need to modify the path.
```

varshaprasad96 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -104,3 +103,92 @@ single-group projects).
+`path: tutorial.kubebuilder.io/project/api/batch/v1beta1`
+
+
+In this way, if the project is not new and has previous APIs already implemented will be in the previous structure. 
+Notice that with the `multi-group` project the Kind API's files are
+created under `api/<group>/<version>` instead of `api/<version>`. 
+Also, note that the controllers will be created under `internal/controller/<group>` instead of `internal/controller`. 
+That is the reason why we moved the previously generated APIs with the provided scripts in the previous steps. 
+Remember to update the references afterwards.
```
```suggestion
In this process, if the project is not new and has previous implemented APIs they would still need to be modified as needed. 
Notice that with the `multi-group` project the Kind API's files are created under `api/<group>/<version>` instead of `api/<version>`. 
Also, note that the controllers will be created under `internal/controller/<group>` instead of `internal/controller`.
 
That is the reason why we moved the previously generated APIs to their respective locations in the new structure. 
Remember to update the references in imports accordingly.
```

varshaprasad96 commented:
/ok-to-test

juanluisvaladas commented:
Thanks for the review!
There were a few trailing whitespaces so I removed those and then I did a manual merge and squash.

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -9,7 +9,7 @@ the Kubebuilder v2 scaffolding (as of Kubebuilder v2.0.0).
 
```
Could we ensure that we explain all with go/v4, which is the current default scaffold and then, at the bottom, we create a new section for the go/v3 layout? Also, could we add a WARNING note saying the go/v3 is deprecated?

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
````diff
@@ -104,3 +103,92 @@ single-group projects).
+mv internal/controller/* internal/controller/batch/
+```
+
+
````
remove the unnecessary spaces

camilamacedo86 reviewed (changes requested):
IHMO, we need to prioritize explaining all using go/v4 (current layout) and then add a note to allow users using go/v3 still know what they should do in their case.

juanluisvaladas commented:
> So, could we not explain it with go/v3 and create a note to let the users know if they are using go/v3, which is deprecated, the steps are mainly the same? However, beware that the layout from go/v3 to go/v4 and the controllers, in this case, are in the root dir and not under the internal directory.

I don't think that's a good idea, when I was doing the migration I found the statements confusing (even if the path was wrong). I think it's better to have duplicate instructions as the other option makes instructions less clear.

I modified this to have go/v4 first and removed the white lines and if you don't find this convincing then I'll remove the Go/v3 section and add the note.

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -21,15 +21,102 @@ to modify the default project structure to support it.
+The instructions vary depending on the project layout. You can verify the version by looking at
+the PROJECT file.
+
+The layout go/v3 is deprecated, if you are usign go/v3 it is recommended that you migrate to go/v4.
```
```suggestion
The layout go/v3 is **deprecated**, if you are using go/v3 it is recommended that you migrate to go/v4. [(More Info)][migration-guide]
```
Can we add the alias for the migration-guide at the bottom.

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -21,15 +21,102 @@ to modify the default project structure to support it.
+[multi-group-issue]: https://github.com/kubernetes-sigs/kubebuilder/issues/923 "Kubebuilder Issue #923"
+[cronjob-tutorial]: /cronjob-tutorial/cronjob-tutorial.md "Tutorial: Building CronJob"
+
+<h1>Using Go/v3</h1>
```
```suggestion
### Using Go/v3
```

camilamacedo86 reviewed (commented):
Just a few nits. Otherwise, it has my /lgtm

juanluisvaladas commented:
Hi, all the issues are addressed, also squashed everything into one commit.

juanluisvaladas commented:
Hi @camilamacedo86, I didn't know this stuff about the project layout go/v2, so I decided to do a couple more modifications.

I rewrote the two notes on top almost from scratch. Please pay extra attention to those two points.
I removed the instructions for go/v3 (which were for go/v2) entirely and just duplicated the step for the controllers.

The rest remains unaltered.

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -4,32 +4,35 @@
-path is now under the `pkg` directory instead. So, you need ensure that you update the
-paths accordingly.
+The layout go/v3 is **deprecated**, if you are using go/v3 it is recommended that you migrate to
+go/v4, however this documentation is still valid. [(More Info)][migration-guide].
```
Just a nit it is missing add the alias at the bottom, see in the preview: https://deploy-preview-3403--kubebuilder.netlify.app/migration/multi-group.html
![Screenshot 2023-05-16 at 22 04 57](https://github.com/kubernetes-sigs/kubebuilder/assets/7708031/3fc1ad3e-dac3-4cdc-b735-d4679e6561f6)

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -4,32 +4,35 @@
+The layout go/v3 is **deprecated**, if you are using go/v3 it is recommended that you migrate to
+go/v4, however this documentation is still valid. [Migration from go/v3 to go/v4][migration-guide].
+
+This document is incompatible with go/v2. The documentation for go/v2 is still available in
+the [legacy documentation](https://book-v2.book.kubebuilder.io/migration/multi-group.html).
+
+The layout go/v1 is incompatible with multi-group.
```
```suggestion
```

It is very unlike someone been using v1/v2 if so they should check the old docs and not this one,
Anyway, if that does not clarify does we can address via chart or issues.

However, here I think we could try to make it more compreensive and help 90% of the audience.

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
````diff
@@ -38,58 +41,72 @@ can check `api/v1/groupversion_info.go` to find that out:
 ```bash
-mkdir apis/batch
-mv api/* apis/batch
-# After ensuring that all was moved successfully remove the old directory `api/`
-rm -rf api/ 
+mv api/* api/batch
````
All steps still required rifght?
We need to create the dir with the group name
mv all api files to each and then remove api/

juanluisvaladas replied:
We only need the mkdir, which I added back. The `mv api/* api/batch` will throw a warning that looks like `mv: cannot move 'batch' to a subdirectory of itself, 'batch/batch'`
 but that can be ignored safely.

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
````diff
@@ -38,58 +41,72 @@ can check `api/v1/groupversion_info.go` to find that out:
+mv internal/controller/* internal/controller/batch/
 ```
 
+For go/v3:
+```bash
+mkdir controller/batch
+mv controller/* controller/batch/
+```
````
````suggestion

<aside class="note">
<h1>If you are using the deprecated layout go/v3</h1>
Then, your layout has not the internal directory. So, you will move the controller(s) under a directory with the name of the API group which it is responsible for manage.
```bash
mkdir controller/batch
mv controller/* controller/batch/
```
````

camilamacedo86 reviewed (commented):
I could just find some nits the rest shows fine for me.

camilamacedo86 commented on docs/book/src/migration/multi-group.md:
```diff
@@ -4,32 +4,35 @@
-While Kubebuilder v2 will not scaffold out a project structure compatible
+While Kubebuilder will not scaffold out a project structure compatible
 with multiple API groups in the same repository by default, it's possible
 to modify the default project structure to support it.
```
```suggestion
to modify the default project structure to support it.

Note that the process mainly is to ensure that your  API(s) and controller(s) will be moved under new directories with their respective group name. 
```

camilamacedo86 replied:
What about we just clarifies as above mainly what we does in the process?

juanluisvaladas commented:
I believe this time everything is addressed :)

camilamacedo86 reviewed (approved):
It shows greta for me 

Thank you a lot for your contribution 🥇

camilamacedo86 commented:
/pull-kubebuilder-test

camilamacedo86 commented:
Hi @juanluisvaladas,

Could you please rebase with master and squash the commits? So that we can get this one merged?

camilamacedo86 reviewed (approved):
/lgtm
